File: generate_gantt.py
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
from matplotlib.patches import FancyBboxPatch
import matplotlib.dates as mdates
from datetime import datetime, timedelta
import numpy as np

# ── Task data ─────────────────────────────────────────────────────────────────
tasks = [
    ("EAS-1  Research & Planning",   datetime(2025, 11,  1), 28),
    ("EAS-2  System Design",         datetime(2025, 11, 29), 21),
    ("EAS-3  Frontend Development",  datetime(2025, 12, 20), 21),
    ("EAS-4  Backend Development",   datetime(2026,  1, 10), 28),
    ("EAS-5  Testing",               datetime(2026,  2,  7), 14),
    ("EAS-6  Implementation",        datetime(2026,  2, 21),  7),
]

# ── Figure setup ──────────────────────────────────────────────────────────────
fig, ax = plt.subplots(figsize=(14, 5))
fig.patch.set_facecolor('white')
ax.set_facecolor('white')

bar_color  = '#5a7192'
grid_color = '#e0e3e8'
text_color = '#2c2c2c'
row_alt    = '#f7f8fa'

n = len(tasks)

# ── Alternating row backgrounds ───────────────────────────────────────────────
for i in range(n):
    if i % 2 == 1:
        ax.axhspan(i - 0.5, i + 0.5, color=row_alt, zorder=0)

# ── Draw bars ─────────────────────────────────────────────────────────────────
for i, (name, start, dur) in enumerate(tasks):
    end = start + timedelta(days=dur)
    ax.barh(
        i, (end - start).days, left=mdates.date2num(start),
        height=0.45, color=bar_color,
        align='center', zorder=3,
        linewidth=0
    )
    # Duration label inside bar
    mid = mdates.date2num(start) + (end - start).days / 2
    label = f"{dur // 7}w"
    ax.text(mid, i, label, ha='center', va='center',
            color='white', fontsize=8.5, fontweight='bold', zorder=4)

# ── X-axis: Nov 2025 – Apr 2026 ───────────────────────────────────────────────
x_start = datetime(2025, 11, 1)
x_end   = datetime(2026,  5, 1)
ax.set_xlim(mdates.date2num(x_start), mdates.date2num(x_end))

# Month major ticks
ax.xaxis.set_major_locator(mdates.MonthLocator())
ax.xaxis.set_major_formatter(mdates.DateFormatter('%b %Y'))
ax.xaxis.set_minor_locator(mdates.WeekdayLocator(byweekday=0))

plt.setp(ax.get_xticklabels(), rotation=0, ha='center',
         fontsize=9, color=text_color, fontweight='600')

# ── Y-axis: task names ────────────────────────────────────────────────────────
ax.set_yticks(range(n))
ax.set_yticklabels([t[0] for t in tasks], fontsize=9.5, color=text_color)
ax.set_ylim(-0.7, n - 0.3)
ax.invert_yaxis()

# ── Grid lines ────────────────────────────────────────────────────────────────
ax.xaxis.grid(True, which='major', color=grid_color, linewidth=1.2, zorder=1)
ax.xaxis.grid(True, which='minor', color=grid_color, linewidth=0.5,
              linestyle='--', zorder=1)
ax.yaxis.grid(True, color=grid_color, linewidth=0.8, zorder=1)

# ── Spines ────────────────────────────────────────────────────────────────────
for spine in ax.spines.values():
    spine.set_edgecolor(grid_color)

# ── Title ─────────────────────────────────────────────────────────────────────
ax.set_title('EasyEdit — Project Gantt Chart  (Nov 2025 – Apr 2026)',
             fontsize=13, fontweight='bold', color=text_color, pad=14)

# ── Top x-axis label (year grouping) ─────────────────────────────────────────
ax.tick_params(axis='x', which='both', top=False)
ax.tick_params(axis='y', which='both', left=False)

plt.tight_layout()
out = r'C:\Users\sauga\OneDrive\Desktop\photo_editor\gantt_chart_final.png'
plt.savefig(out, dpi=180, bbox_inches='tight', facecolor='white')
print(f"Saved → {out}")

File: generate_gantt.js
const { createCanvas } = require('canvas');
const fs = require('fs');
const path = require('path');

// ── Task data ──────────────────────────────────────────────────────────────
const tasks = [
    { name: '☑  EAS-1  Research & Planning', start: new Date('2025-11-01'), weeks: 4 },
    { name: '☑  EAS-2  System Design', start: new Date('2025-11-29'), weeks: 3 },
    { name: '☑  EAS-3  Frontend Development', start: new Date('2025-12-20'), weeks: 3 },
    { name: '☑  EAS-4  Backend Development', start: new Date('2026-01-10'), weeks: 4 },
    { name: '☑  EAS-5  Testing', start: new Date('2026-02-07'), weeks: 2 },
    { name: '☑  EAS-6  Implementation', start: new Date('2026-02-21'), weeks: 1 },
];

// ── Chart ranges ───────────────────────────────────────────────────────────
const rangeStart = new Date('2025-11-01');
const rangeEnd = new Date('2026-05-01');
const totalDays = (rangeEnd - rangeStart) / 86400000;

const months = [
    { label: 'Nov 2025', date: new Date('2025-11-01') },
    { label: 'Dec 2025', date: new Date('2025-12-01') },
    { label: 'Jan 2026', date: new Date('2026-01-01') },
    { label: 'Feb 2026', date: new Date('2026-02-01') },
    { label: 'Mar 2026', date: new Date('2026-03-01') },
    { label: 'Apr 2026', date: new Date('2026-04-01') },
];

// ── Canvas dimensions ──────────────────────────────────────────────────────
const W = 1200;
const H = 380;
const LEFT_PAD = 230;  // task name column
const TOP_PAD = 80;   // header
const BOT_PAD = 40;
const chartW = W - LEFT_PAD - 20;
const chartH = H - TOP_PAD - BOT_PAD;
const rowH = chartH / tasks.length;

const canvas = createCanvas(W, H);
const ctx = canvas.getContext('2d');

// ── Background ─────────────────────────────────────────────────────────────
ctx.fillStyle = '#ffffff';
ctx.fillRect(0, 0, W, H);

// helper: day offset → x pixel
function dayToX(date) {
    const days = (date - rangeStart) / 86400000;
    return LEFT_PAD + (days / totalDays) * chartW;
}

// ── Alternating row shading ────────────────────────────────────────────────
tasks.forEach((_, i) => {
    const y = TOP_PAD + i * rowH;
    ctx.fillStyle = i % 2 === 0 ? '#ffffff' : '#f5f7fa';
    ctx.fillRect(LEFT_PAD, y, chartW, rowH);
});

// ── Vertical month dividers ────────────────────────────────────────────────
ctx.strokeStyle = '#d4d8de';
ctx.lineWidth = 1;
months.forEach(m => {
    const x = dayToX(m.date);
    ctx.beginPath();
    ctx.moveTo(x, TOP_PAD);
    ctx.lineTo(x, H - BOT_PAD);
    ctx.stroke();
});

// ── Horizontal row dividers ────────────────────────────────────────────────
for (let i = 0; i <= tasks.length; i++) {
    const y = TOP_PAD + i * rowH;
    ctx.beginPath();
    ctx.moveTo(LEFT_PAD, y);
    ctx.lineTo(W - 20, y);
    ctx.stroke();
}

// ── Header background ──────────────────────────────────────────────────────
ctx.fillStyle = '#edf0f5';
ctx.fillRect(0, 0, W, TOP_PAD);

// ── Header: vertical separator line at LEFT_PAD ───────────────────────────
ctx.strokeStyle = '#b8bec9';
ctx.lineWidth = 1.5;
ctx.beginPath();
ctx.moveTo(LEFT_PAD, 0);
ctx.lineTo(LEFT_PAD, H - BOT_PAD);
ctx.stroke();

// ── Header month labels ────────────────────────────────────────────────────
ctx.fillStyle = '#3a4a5c';
ctx.font = 'bold 13px Arial';
ctx.textAlign = 'center';

for (let i = 0; i < months.length; i++) {
    const mStart = months[i].date;
    const mEnd = i + 1 < months.length ? months[i + 1].date : rangeEnd;
    const xLeft = dayToX(mStart);
    const xRight = dayToX(mEnd);
    const xMid = (xLeft + xRight) / 2;

    ctx.fillStyle = '#3a4a5c';
    ctx.fillText(months[i].label, xMid, TOP_PAD - 10);

    // light vertical inside header
    if (i > 0) {
        ctx.strokeStyle = '#c8cdd8';
        ctx.lineWidth = 1;
        ctx.beginPath();
        ctx.moveTo(xLeft, 10);
        ctx.lineTo(xLeft, TOP_PAD - 4);
        ctx.stroke();
    }
}

// ── Header "Items" label ───────────────────────────────────────────────────
ctx.fillStyle = '#3a4a5c';
ctx.font = 'bold 13px Arial';
ctx.textAlign = 'center';
ctx.fillText('Items', LEFT_PAD / 2, TOP_PAD - 10);

// ── Task name labels ───────────────────────────────────────────────────────
ctx.font = '12px Arial';
ctx.textAlign = 'left';
tasks.forEach((t, i) => {
    const y = TOP_PAD + i * rowH + rowH / 2 + 4;
    ctx.fillStyle = '#2c3e50';
    ctx.fillText(t.name, 12, y);
});

// ── Gantt bars ─────────────────────────────────────────────────────────────
const BAR_COLOR = '#5a7192';
const LABEL_COLOR = '#ffffff';
const BAR_RADIUS = 4;
const barPad = rowH * 0.22;

tasks.forEach((t, i) => {
    const endDate = new Date(t.start);
    endDate.setDate(endDate.getDate() + t.weeks * 7);

    const x = dayToX(t.start);
    const x2 = dayToX(endDate);
    const bW = x2 - x;
    const y = TOP_PAD + i * rowH + barPad;
    const bH = rowH - barPad * 2;

    // Rounded rectangle bar
    ctx.fillStyle = BAR_COLOR;
    ctx.beginPath();
    ctx.moveTo(x + BAR_RADIUS, y);
    ctx.lineTo(x + bW - BAR_RADIUS, y);
    ctx.quadraticCurveTo(x + bW, y, x + bW, y + BAR_RADIUS);
    ctx.lineTo(x + bW, y + bH - BAR_RADIUS);
    ctx.quadraticCurveTo(x + bW, y + bH, x + bW - BAR_RADIUS, y + bH);
    ctx.lineTo(x + BAR_RADIUS, y + bH);
    ctx.quadraticCurveTo(x, y + bH, x, y + bH - BAR_RADIUS);
    ctx.lineTo(x, y + BAR_RADIUS);
    ctx.quadraticCurveTo(x, y, x + BAR_RADIUS, y);
    ctx.closePath();
    ctx.fill();

    // Duration label
    ctx.fillStyle = LABEL_COLOR;
    ctx.font = 'bold 11px Arial';
    ctx.textAlign = 'center';
    ctx.fillText(`${t.weeks}w`, x + bW / 2, y + bH / 2 + 4);
});

// ── Title ──────────────────────────────────────────────────────────────────
ctx.fillStyle = '#1a2535';
ctx.font = 'bold 15px Arial';
ctx.textAlign = 'left';
ctx.fillText('EasyEdit AI Photo Editor — Project Gantt Chart (Nov 2025 – Apr 2026)', 12, 24);

// ── Bottom border ──────────────────────────────────────────────────────────
ctx.strokeStyle = '#d4d8de';
ctx.lineWidth = 1;
ctx.beginPath();
ctx.moveTo(0, H - BOT_PAD);
ctx.lineTo(W, H - BOT_PAD);
ctx.stroke();

// ── Legend ─────────────────────────────────────────────────────────────────
const legendY = H - BOT_PAD + 12;
ctx.fillStyle = BAR_COLOR;
ctx.fillRect(LEFT_PAD, legendY, 16, 11);
ctx.fillStyle = '#555';
ctx.font = '11px Arial';
ctx.textAlign = 'left';
ctx.fillText('Project Phase  (bar width = duration)', LEFT_PAD + 22, legendY + 10);

// ── Save ───────────────────────────────────────────────────────────────────
const out = path.join('C:\\Users\\sauga\\OneDrive\\Desktop\\photo_editor', 'gantt_chart_final.png');
const buf = canvas.toBuffer('image/png');
fs.writeFileSync(out, buf);
console.log('Saved →', out);
